src/helpers/test/RenderContexts.tsx

```diff
@@ -0,0 +1,62 @@
+/*
+MIT License
+
+Copyright (c) 2021-present, Elastic NV
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+import React from "react";
+
+export interface ContextRenderer<ContextValue> {
+  defaults: ContextValue;
+  Context: React.Context<ContextValue>;
+  overrides?: Partial<ContextValue>;
+}
+
+/**
+ * Takes a list of React Contexts and values and recursively renders them overtop of a given component.
+ * @returns Rendered contexts wrapping child component.
+ */
+export const RenderContexts = ({
+  contexts,
+  children,
+}: {
+  /* The below line circumvents TS screaming about the different context types not being
+   * compatible with a union of all types, i.e. "IRecordingContext is incompatible with IRecordingContext | IStepsContext".
+   *
+   * Given the purpose of this function is to recursively render a generic list of contexts, it's not
+   * imperative to know or care what the type of the context is.
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  contexts: Array<ContextRenderer<any>>;
+  children?: React.ReactNode;
+}) => {
+  const context = contexts.shift();
+  if (typeof context === "undefined") {
+    return <>{children}</>;
+  }
+  const { defaults, Context, overrides } = context;
+  const value = overrides ? { ...defaults, ...overrides } : defaults;
+  return (
+    <Context.Provider value={value}>
+      <RenderContexts contexts={contexts}>{children}</RenderContexts>
```

Comment:
Really elegant recursive implementation, I love it.

Reply:
Ah thanks, I was concerned it was overly clever, but my hope is that it will work seamlessly enough that no one will ever have to look at it.